staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition_test.go

```diff
@@ -0,0 +1,172 @@
+	"k8s.io/apiserver/pkg/admission"
+	celconfig "k8s.io/apiserver/pkg/apis/cel"
+	"k8s.io/apiserver/pkg/cel/environment"
+)
+
+type testVariable struct {
+	name       string
+	expression string
+}
+
+func (t *testVariable) GetExpression() string {
+	return t.expression
+}
+
+func (t *testVariable) ReturnTypes() []*cel.Type {
+	return []*cel.Type{cel.AnyType}
+}
+
+func (t *testVariable) GetName() string {
+	return t.name
+}
+
+func TestCompositedPolicies(t *testing.T) {
+	cases := []struct {
+		name                 string
+		variables            []NamedExpressionAccessor
+		expression           string
+		attributes           admission.Attributes
+		expectedResult       any
+		expectErr            bool
+		expectedErrorMessage string
+		runtimeCostBudget    int64
+	}{
+		{
+			name: "simple",
+			variables: []NamedExpressionAccessor{
+				&testVariable{
+					name:       "name",
+					expression: "object.metadata.name",
+				},
+			},
+			attributes:     endpointCreateAttributes(),
+			expression:     "variables.name == 'endpoints1'",
+			expectedResult: true,
+		},
+		{
+			name: "delayed compile error",
+			variables: []NamedExpressionAccessor{
+				&testVariable{
+					name:       "name",
+					expression: "1 == '1'", // won't compile
+				},
+			},
+			attributes:           endpointCreateAttributes(),
+			expression:           "variables.name == 'endpoints1'",
+			expectErr:            true,
+			expectedErrorMessage: `composited variable "name" fails to compile:`,
+		},
+		{
+			name: "delayed eval error",
+			variables: []NamedExpressionAccessor{
+				&testVariable{
+					name:       "name",
+					expression: "object.spec.subsets[114514].addresses.size()", // array index out of bound
+				},
+			},
+			attributes:           endpointCreateAttributes(),
+			expression:           "variables.name == 'endpoints1'",
+			expectErr:            true,
+			expectedErrorMessage: `composited variable "name" fails to evaluate:`,
+		},
+		{
+			name: "out of budget during lazy evaluation",
+			variables: []NamedExpressionAccessor{
+				&testVariable{
+					name:       "name",
+					expression: "object.metadata.name", // cost = 3
+				},
+			},
+			attributes:           endpointCreateAttributes(),
+			expression:           "variables.name == 'endpoints1'", // cost = 3
+			expectedResult:       true,
+			runtimeCostBudget:    4, // enough for main variable but not for entire expression
+			expectErr:            true,
+			expectedErrorMessage: "running out of cost budget",
+		},
+		{
+			name: "lazy evaluation, budget counts only once",
+			variables: []NamedExpressionAccessor{
+				&testVariable{
+					name:       "name",
+					expression: "object.metadata.name", // cost = 3
+				},
+			},
+			attributes:        endpointCreateAttributes(),
+			expression:        "variables.name == 'endpoints1' && variables.name == 'endpoints1' ", // cost = 7
+			expectedResult:    true,
+			runtimeCostBudget: 10, // enough for one lazy evaluation but not two, should pass
+		},
```

Comment:
Missing tests for:
- multiple variables expr
- multiple variables operation
- Do we allow variable nesting?

Reply:
These are tested in the E2E suite.

Reply:
Let me enhance it a bit to use multiple variables